src/connectors/impls/s3_source.rs

```diff
@@ -0,0 +1,320 @@
+
+            let handle = task::spawn(fetch_object_task(
+                task_client,
+                task_bucket,
+                task_rx,
+                task_sender,
+            ));
+            self.handles.push(handle);
+        }
+
+        // Key Fetcher Task
+        let bucket = self.config.bucket.clone();
+        let prefix = self.config.prefix.clone();
+        let delim = self.config.delimiter.clone();
+        task::spawn(fetch_keys_task(client, bucket, prefix, delim, tx_key));
+
+        Ok(true)
+    }
+
+    fn default_codec(&self) -> &str {
+        "json"
+    }
+}
+
+async fn fetch_keys_task(
+    client: S3Client,
+    bucket: String,
+    prefix: Option<String>,
+    delim: Option<String>,
+    sender: Sender<KeyPayload>,
+) -> Result<()> {
+    // fetch first pool of keys.
+    let mut resp = client
+        .list_objects_v2()
+        .bucket(bucket.clone())
+        .set_prefix(prefix.clone())
+        .set_delimiter(delim.clone())
+        .send()
+        .await?;
+
+    let mut stream = 0; // for the Channel Source
+    let mut last_key_fetched: Option<String> = None;
+    'outer: loop {
+        match resp.contents.take() {
+            None => {}
+            Some(entries) => {
+                // FIXME: Could there be issues setting this before fetching all the keys
+                match entries.last() {
+                    None => {
+                        // No more keys/objects
+                        break 'outer;
+                    }
+                    Some(obj) => last_key_fetched = obj.key.clone(),
+                }
+
+                for object_data in entries {
+                    sender
+                        .send(KeyPayload {
+                            object_data,
+                            stream,
+                        })
+                        .await?;
+                    stream += 1;
+                }
+            }
+        }
+
+        if !resp.is_truncated {
+            // No more entries.
+            break 'outer;
+        }
+
+        resp = client
+            .list_objects_v2()
+            .bucket(bucket.clone())
+            .set_prefix(prefix.clone())
+            .set_delimiter(delim.clone())
+            .set_start_after(last_key_fetched.take())
+            .send()
+            .await?;
+    }
+
+    Ok(())
+}
+
+async fn fetch_object_task(
+    client: S3Client,
+    bucket: String,
+    recvr: Receiver<KeyPayload>,
+    sender: Sender<SourceReply>,
+) -> Result<()> {
+    // gets key from the recvr
+    // fetches the object from s3
+    // send to the ChannelSource channel
+
+    // would make sense to use a preallocated buffer.
+    let mut v = Vec::new();
+
+    while let Ok(KeyPayload {
+        object_data,
+        stream,
+    }) = recvr.recv().await
+    {
+        // Auto handled.
+        // sender.send(SourceReply::StartStream(stream)).await?;
+
+        // Fetch the object here
```

Comment:
could you add a `TODO` for checking the object size and if it crosses a certain threshold, we download part by part either via https://docs.rs/aws-sdk-s3/latest/aws_sdk_s3/client/fluent_builders/struct.GetObject.html#method.set_part_number or via https://docs.rs/aws-sdk-s3/latest/aws_sdk_s3/client/fluent_builders/struct.GetObject.html#method.set_range